src/Avalonia.Visuals/Rect.cs

```diff
@@ -421,11 +421,48 @@ public Rect Translate(Vector offset)
         }
 
         /// <summary>
-        /// Gets the union of two rectangles.
-        /// </summary>
-        /// <param name="rect">The other rectangle.</param>
-        /// <returns>The union.</returns>
-        public Rect Union(Rect rect)
+		/// Normalizes the rectangle so both the <see cref="Width"/> and <see cref="Height"/> are positive, without changing the location of the rectangle
+		/// </summary>
+        /// <returns>Normalized Rect</returns>
+		/// <remarks>
+		/// Empty rect will be return when Rect contains invalid values. Like NaN.
+		/// </remarks>
+		public Rect Normalize()
+        {
+            Rect rect = this;            
+
+            if(double.IsNaN(rect.Right) || double.IsNaN(rect.Bottom) || double.IsNaN(rect.X) || double.IsNaN(rect.Y) || double.IsNaN(Height) || double.IsNaN(Width))
+            {
+                return Rect.Empty;
+            }
+
+            if (rect.Width < 0)
```

Comment:
Why can we construct a Rect with negative width / height? Shouldn't we throw an exception in the ctor if someone tries that?

Reply:
A rect with negative with is really useful for a lot of mathematics, and is supported by just about every graphics library afaik. It simply means a rect that is anchored at the bottom/right instead of top/left. `Rect.Normalize` is a common method for rects:

https://github.com/picoe/Eto/blob/develop/src/Eto/Drawing/RectangleF.cs#L33-L53
https://www.pdftron.com/api/PDFNet/html/M_pdftron_PDF_Rect_Normalize.htm
https://apireference.aspose.com/imaging/net/aspose.imaging/rectangle/methods/normalize

as just some examples that came up in a google.

Previously, `Rect` implicitly supported non-normalized rects (in that it didn't check them) but IMO we should support them properly because they're really useful for certain algorithms. This is the first step, but we also need to normalize rects before carrying out some other operations, e.g. `Rect.Contains`.

Reply:
Actually GDI, simply says "some stuff won't work unless you normalize the rect first". That might be acceptable: https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-ptinrect

Reply:
In skia this is called `Sort`: https://github.com/google/skia/blob/master/include/core/SkRect.h#L527-L539

Reply:
Thanks for your explanations